iceoryx2/src/node/mod.rs

```diff
@@ -408,30 +492,55 @@ impl<Service: service::Service> Node<Service> {
     }
 
     /// Returns a list of [`NodeState`] of all [`Node`]s in the system under a given [`Config`].
-    pub fn list(config: &Config) -> Result<Vec<NodeState<Service>>, NodeListFailure> {
+    ///
+    /// ```
+    /// # use iceoryx2::prelude::*;
+    /// Node::<zero_copy::Service>::list(Config::get_global_config(), |node_state| {
+    ///     println!("found node {:?}", node_state?);
+    ///     Ok(CallbackProgression::Continue)
+    /// });
+    /// ```
+    pub fn list<
+        F: FnMut(
+            Result<NodeState<Service>, NodeListFailure>,
+        ) -> Result<CallbackProgression, NodeListFailure>,
+    >(
+        config: &Config,
+        mut callback: F,
+    ) -> Result<(), NodeListFailure> {
+        let msg = "Unable to iterate over Node list";
+        let origin = "Node::list()";
         let monitoring_config = node_monitoring_config::<Service>(config);
-        let mut nodes = vec![];
 
-        for node_name in &Self::list_all_nodes(&monitoring_config)? {
-            let id_value = core::str::from_utf8(node_name.as_bytes()).unwrap();
-            let id_value = id_value.parse::<u128>().unwrap();
-
-            let details = Self::get_node_details(config, node_name).unwrap_or_default();
-
-            let node_view = AliveNodeView::<Service> {
-                id: id_value.into(),
-                details,
-                _service: PhantomData,
-            };
-
-            match Self::get_node_state(&monitoring_config, node_name)? {
-                State::DoesNotExist => (),
-                State::Alive => nodes.push(NodeState::Alive(node_view)),
-                State::Dead => nodes.push(NodeState::Dead(DeadNodeView(node_view))),
-            };
+        match Self::list_all_nodes(&monitoring_config) {
+            Ok(node_list) => {
+                for node_name in node_list {
+                    let node_id = core::str::from_utf8(node_name.as_bytes()).unwrap();
+                    let node_id = NodeId(node_id.parse::<u128>().unwrap().into());
+
+                    match NodeState::new(&node_id, config) {
+                        Ok(Some(node_state)) => {
+                            if callback(Ok(node_state))? == CallbackProgression::Stop {
```

Comment:
Another reason to split it. It's always used like `callback(Ok(...))` or `callback(Err(...))` and the first thing in the callback is a match on the `Result` input value.